core-bundle/src/Altcha/Altcha.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Altcha;
+
+use Contao\CoreBundle\Altcha\Config\Algorithm;
+use Contao\CoreBundle\Altcha\Exception\InvalidAlgorithmException;
+use Contao\CoreBundle\Entity\Altcha as AltchaEntity;
+use Contao\CoreBundle\Repository\AltchaRepository;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * @internal
+ */
+class Altcha
+{
+    public function __construct(
+        private readonly AltchaRepository $repository,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly string $secret,
+        private readonly string $altchaAlgorithm,
+        private readonly int $altchaRangeMin,
+        private readonly int $altchaRangeMax,
+        private readonly int $altchaChallengeExpiry,
+    ) {
+    }
+
+    /**
+     * @throws InvalidAlgorithmException
+     */
+    public function createChallenge(string|null $salt = null, int|null $number = null): AltchaChallenge
+    {
+        $salt ??= bin2hex(random_bytes(12));
+        $number ??= random_int($this->altchaRangeMin, $this->altchaRangeMax);
+        $algorithms = Algorithm::values();
+
+        if (!\in_array($this->altchaAlgorithm, $algorithms, true)) {
+            $strError = 'Invalid algorithm selected. It has to be set to one of these "%s".';
+
+            throw new InvalidAlgorithmException(sprintf($strError, implode('", "', $algorithms)));
+        }
+
+        $algorithm = str_replace('-', '', strtolower($this->altchaAlgorithm));
+
+        // Generate the challenge
+        $challenge = hash($algorithm, $salt.$number);
+
+        // Generate the signature
+        $signature = hash_hmac($algorithm, $challenge, $this->secret);
+
+        // The challenge expires in 1 hour (default) and it must be saved in the database
+        // to prevent replay attacks.
+        $expiryDate = date('Y-m-d\\TH:i:sP', time() + $this->altchaChallengeExpiry);
+        $objAltcha = new AltchaEntity($challenge, new \DateTimeImmutable($expiryDate));
```

Comment:
I think this cannot work because that class does not exist (that's why we need tests ;)).